nifi-mock/src/main/java/org/apache/nifi/util/MockPropertyValue.java

```diff
@@ -202,6 +203,9 @@ public PropertyValue evaluateAttributeExpressions(final AttributeValueDecorator
 
     @Override
     public PropertyValue evaluateAttributeExpressions(final FlowFile flowFile) throws ProcessException {
+        if (flowFile == null) {
```

Comment:
This feels a little bit odd to me, as it's very unclear just from looking at this method what the intent is here - i think it would be cleaner to pass down a boolean argument that indicates whether or not attributes are available. It would be `true` in all cases except for `evaluateAttributeExpressions()` with no arguments and `evaluateAttributeExpressions(AttributeValueDecorator decorator)` - or otherwise to just have the no-arg version call the one that takes only the decorator and implement that method there instead of calling another override of the method. The actual implementation is only about 4-5 lines of code and quite trivial anyway.

Reply:
> This feels a little bit odd to me, as it's very unclear just from looking at this method what the intent is here

The intent is to mirror the behavior of running NiFi when you pass a null flowfile object into `evaluateExpressionLanguage(FlowFile)`. This works on a running processor:

```
FlowFile parent = null;
String charset = context.getProperty(CHARSET).evaluateAttributeExpressions(parent).getValue();
```

it does not work against the testing framework. So it seems natural to me that the testing framework just handle a null flowfile as an empty set of attributes the way live NiFi does.

Reply:
Sorry - don't think I was clear in what I was trying to communicate here. I do understand the intent. But it took me a while to understand how the PR differed logically from what already was there. What would seem natural to me is to simply pass the null FlowFile along to an override, filling in other default values, as it already does. But clearly that is a bug because of how this is handled in the actual implementing override. At first look, though, it seemed to me to be the same logic because in many places an empty map is treated the same as a null map.

So while I now understand what it's doing, I think it's a bit confusing at first glance. That's why i suggested passing a boolean down the stack. The PR wouldn't cause me much angst if there were just some inline comments, I think, about how the implementing override checks if flowfile is null and variables is null and if so handles things differently... but at that point it makes it really easy to change the implementation later and leave comments in the code that are no longer valid. but i'd be okay either way.

Reply:
Ok. I'll comment it up and make it clear what it's doing so anyone jumping in there later will understand exactly what's going on there.